umbraco/Umbraco-CMS
# V9: Fix ImageSharp integration and add support for custom crops

This fixes an issue where custom crops was not reflected if you zoomed in or moved the crop around.

This was because we had replace `ImageProcessor` with `ImageSharp.Web`, which uses different parameters and doesn't support the type of cropping we use out of box.

To fix this I created a custom Crop `IImageWebProcessor` which grabs the old crop parameter from the URL string, and recalculate it into the ImageSharp `Rectangle` format, and then crop the image, which will be sent on the resizer where it will be resized. 

I also considered creating a custom resizer that uses the TargetRectagle and the manual mode, however, I think this will be much harder to manage and maintain, than separating the concerns of cropping and resizing. The documentation is not entirely clear if it is at all possible to do it in this manner regardless.


### Testing

To this out, you must 
1. Create a document type that uses the media picker
2. Specify a crop on the picker
3. Render the media in your template
3. Create a piece of content with an image in the media picker
   1. Specify your crop on the image, to see the difference you must move the crop and/or use the slider to resize it
4. Navigate to the page and see the different pre/post PR 

(You might have to clear your browser cache) 

I made an image with a custom crop, before the changes it looked like:

![before](https://user-images.githubusercontent.com/16456704/124932179-237f4880-e003-11eb-93eb-45e3f241d69b.png)

After the changes:

![after](https://user-images.githubusercontent.com/16456704/124932219-2c701a00-e003-11eb-8684-51b49a543457.png)
